packages/affiliates/libs/uniswap/models.ts

```diff
@@ -0,0 +1,253 @@
+  tickUpper: string;
+  token0?: string;
+  token1?: string;
+  liquidity?: string;
+  feeGrowthInside0LastX128?: string;
+  feeGrowthInside1LastX128?: string;
+  tokensOwed0?: string;
+  tokensOwed1?: string;
+  pool?: string;
+};
+
+// These positions are from core pool contracts and are indexed using hash of upper/lower/user address.
+export const Positions = (config: any, table: ReturnType<typeof Cache>) => {
+  function makeId(data: Position) {
+    return getPositionKey(data.operator, data.tickLower, data.tickUpper);
+  }
+  async function create(data: Position) {
+    const id = makeId(data);
+    assert(!(await has(id)), "Position exists");
+    return set({ id, ...data });
+  }
+  async function set(data: Position) {
+    assert(data.id, "Position requires id");
+    await table.set(data.id, { ...data });
+    return data;
+  }
+  async function get(id: string) {
+    assert(await table.has(id), "Position does not exist");
+    return table.get(id);
+  }
+  async function has(id: string) {
+    return table.has(id);
+  }
+  async function forEach(cb: (value: Position, index: number, array: Position[]) => void) {
+    const list: Position[] = (await table.list()) as Position[];
+    list.forEach(cb);
+  }
+  async function update(id: string, data: any = {}) {
+    const position = await get(id);
+    return set({ ...position, ...data });
+  }
+
+  return {
+    create,
+    set,
+    get,
+    has,
+    update,
+    forEach,
+    list: table.list
+  };
+};
+
+export type GlobalState = {
+  id?: string;
+  token0: string;
+  token1: string;
+  fee: string;
+  sqrtPriceX96?: string;
+  tick?: number;
+  liquidity?: string;
+  feeProtocol?: number;
+  feeGrowthGlobal0X128?: string;
+  feeGrowthGlobal1X128?: string;
+  protocolFeesToken0?: string;
+  protocolFeesToken1?: string;
+  address?: string;
+};
+// pool id is tokenA, tokenB, fee
+export const Pools = (config: any, table: ReturnType<typeof Cache>) => {
+  function makeId(state: GlobalState) {
+    const { token0, token1, fee } = state;
+    if (token0 < token1) {
+      return [token0, token1, fee].join("!");
+    }
+    return [token1, token0, fee].join("!");
+  }
+  async function create(state: GlobalState) {
+    const id = makeId(state);
+    assert(!(await has(id)), "State already exists:" + id);
+    return set({ id, ...state });
+  }
+  async function set(state: GlobalState) {
+    assert(state.id, "requires id");
+    await table.set(state.id, state);
+    return state;
+  }
+  async function get(id: string) {
+    assert(await has(id), "No such pool state");
+    return table.get(id);
+  }
+  async function has(id: string) {
+    return table.has(id);
+  }
+  async function update(id: string, data: any = {}) {
+    const pool = await get(id);
+    return set({ ...pool, ...data });
+  }
+  async function list() {
+    return table.list();
+  }
+  return {
+    create,
+    set,
+    get,
+    has,
+    update,
+    list
+  };
+};
+
+export type TickInfo = {
```

Comment:
the main concern with allowing BNs in here is that its not compatible between web3 numbers and ethers afaik.  slightly less important is that you cant serialize BigNumbers or BNs to a database and while not needed for this, you never know when that could be handy.  You are right it would be more performant to just choose one and go with it though since everything is in memory.